File: Sources/ExcalidrawGeometry/ElementGeometry.swift
import ExcalidrawMath
import ExcalidrawModel
import Foundation

/// Element-aware geometry: absolute coordinates, bounds (rotation-aware), and
/// outline extraction. Ports the bounds logic from
/// `packages/element/src/bounds.ts`.
///
/// Scope note: linear/arrow bounds and outlines use the element's point
/// polyline directly. For straight segments this is exact; for curved
/// (rounded) arrows it is the control-polygon approximation — exact cubic-Bézier
/// bounds arrive with the rough.js shape port (later phase).
public enum ElementGeometry {
    /// `[x1, y1, x2, y2, cx, cy]` in scene coordinates (`getElementAbsoluteCoords`).
    public static func absoluteCoords(_ element: ExcalidrawElement) -> (
        x1: Double, y1: Double, x2: Double, y2: Double, cx: Double, cy: Double
    ) {
        let base = element.base
        if let pts = scenePoints(element), !pts.isEmpty,
           case let kind = element.kind, isPolylineKind(kind) {
            let box = BoundingBox(points: pts)!
            return (box.minX, box.minY, box.maxX, box.maxY,
                    (box.minX + box.maxX) / 2, (box.minY + box.maxY) / 2)
        }
        let x1 = base.x, y1 = base.y
        let x2 = base.x + base.width, y2 = base.y + base.height
        return (x1, y1, x2, y2, (x1 + x2) / 2, (y1 + y2) / 2)
    }

    /// Bounding box of the element. When `nonRotated` is true the element's own
    /// rotation is ignored (`getElementBounds` / `ElementBounds.calculateBounds`).
    public static func bounds(_ element: ExcalidrawElement, nonRotated: Bool = false) -> BoundingBox {
        let (x1, y1, x2, y2, cx, cy) = absoluteCoords(element)
        if nonRotated || element.base.angle == 0 {
            return BoundingBox(minX: x1, minY: y1, maxX: x2, maxY: y2)
        }
        let angle = element.base.angle
        let center = Point(cx, cy)

        switch element.kind {
        case .freedraw, .line, .arrow:
            let rotated = (scenePoints(element) ?? []).map { $0.rotated(around: center, by: angle) }
            return BoundingBox(points: rotated) ?? BoundingBox(minX: x1, minY: y1, maxX: x2, maxY: y2)
        case .ellipse:
            let w = (x2 - x1) / 2
            let h = (y2 - y1) / 2
            let ww = hypot(w * cos(angle), h * sin(angle))
            let hh = hypot(h * cos(angle), w * sin(angle))
            return BoundingBox(minX: cx - ww, minY: cy - hh, maxX: cx + ww, maxY: cy + hh)
        default:
            let corners = unrotatedOutline(element).points.map { $0.rotated(around: center, by: angle) }
            return BoundingBox(points: corners) ?? BoundingBox(minX: x1, minY: y1, maxX: x2, maxY: y2)
        }
    }

    /// Combined bounds of several elements (`getCommonBounds`). `nil` if empty.
    public static func commonBounds(_ elements: [ExcalidrawElement]) -> BoundingBox? {
        guard !elements.isEmpty else { return nil }
        return elements.reduce(nil) { acc, element in
            let b = bounds(element)
            return acc.map { $0.union(b) } ?? b
        }
    }

    // MARK: Outline

    /// An element's outline as unrotated scene-space vertices, plus whether the
    /// path is closed (forms an interior).
    struct Outline {
        var points: [Point]
        var closed: Bool
    }

    static func unrotatedOutline(_ element: ExcalidrawElement) -> Outline {
        let base = element.base
        let x1 = base.x, y1 = base.y
        let x2 = base.x + base.width, y2 = base.y + base.height
        let cx = (x1 + x2) / 2, cy = (y1 + y2) / 2

        switch element.kind {
        case .diamond:
            return Outline(points: [Point(cx, y1), Point(x2, cy), Point(cx, y2), Point(x1, cy)], closed: true)
        case let .line(props):
            let pts = props.points.map { Point(base.x + $0.x, base.y + $0.y) }
            return Outline(points: pts, closed: props.polygon || isPathALoop(props.points))
        case .arrow:
            let pts = (scenePoints(element) ?? [])
            return Outline(points: pts, closed: false)
        case let .freedraw(props):
            let pts = props.points.map { Point(base.x + $0.x, base.y + $0.y) }
            return Outline(points: pts, closed: isPathALoop(props.points))
        case .ellipse:
            // Ellipses are handled analytically, not as a polygon.
            return Outline(points: [Point(x1, y1), Point(x2, y1), Point(x2, y2), Point(x1, y2)], closed: true)
        default:
            return Outline(points: [Point(x1, y1), Point(x2, y1), Point(x2, y2), Point(x1, y2)], closed: true)
        }
    }

    // MARK: Helpers

    /// Whether a path of linear points is closed (`isPathALoop`): ≥3 points and
    /// endpoints within `LINE_CONFIRM_THRESHOLD`.
    static func isPathALoop(_ points: [Point]) -> Bool {
        guard points.count >= 3, let first = points.first, let last = points.last else { return false }
        return first.distance(to: last) <= Constants.lineConfirmThreshold
    }

    /// Scene-space points for polyline elements (freedraw/line/arrow), else nil.
    static func scenePoints(_ element: ExcalidrawElement) -> [Point]? {
        let base = element.base
        switch element.kind {
        case let .freedraw(props): return props.points.map { Point(base.x + $0.x, base.y + $0.y) }
        case let .line(props): return props.points.map { Point(base.x + $0.x, base.y + $0.y) }
        case let .arrow(props): return props.points.map { Point(base.x + $0.x, base.y + $0.y) }
        default: return nil
        }
    }

    private static func isPolylineKind(_ kind: ElementKind) -> Bool {
        switch kind {
        case .freedraw, .line, .arrow: return true
        default: return false
        }
    }

    public enum Constants {
        public static let lineConfirmThreshold = 40.0
    }
}

File: web/packages/geometry/src/element-geometry.ts
import { Point } from "@cyberdynecorpai/math";
import type { ExcalidrawElement, LocalPoint } from "@cyberdynecorpai/model";
import { BoundingBox } from "./bounding-box.js";

export const LINE_CONFIRM_THRESHOLD = 40;

type PolylineElement = Extract<ExcalidrawElement, { type: "freedraw" | "line" | "arrow" }>;

function isPolyline(el: ExcalidrawElement): el is PolylineElement {
  return el.type === "freedraw" || el.type === "line" || el.type === "arrow";
}

/** Scene-space points for polyline elements (freedraw/line/arrow), else null. */
export function scenePoints(el: ExcalidrawElement): Point[] | null {
  if (!isPolyline(el)) return null;
  return el.points.map((p: LocalPoint) => new Point(el.x + p[0], el.y + p[1]));
}

/** Whether a path of relative points is closed (`isPathALoop`). */
export function isPathALoop(points: LocalPoint[]): boolean {
  if (points.length < 3) return false;
  const first = points[0]!;
  const last = points[points.length - 1]!;
  return (
    new Point(first[0], first[1]).distance(new Point(last[0], last[1])) <= LINE_CONFIRM_THRESHOLD
  );
}

export interface AbsoluteCoords {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  cx: number;
  cy: number;
}

/** `[x1, y1, x2, y2, cx, cy]` in scene coordinates (`getElementAbsoluteCoords`). */
export function absoluteCoords(el: ExcalidrawElement): AbsoluteCoords {
  const pts = scenePoints(el);
  if (pts !== null && pts.length > 0) {
    const box = BoundingBox.fromPoints(pts)!;
    return {
      x1: box.minX,
      y1: box.minY,
      x2: box.maxX,
      y2: box.maxY,
      cx: (box.minX + box.maxX) / 2,
      cy: (box.minY + box.maxY) / 2,
    };
  }
  const x1 = el.x;
  const y1 = el.y;
  const x2 = el.x + el.width;
  const y2 = el.y + el.height;
  return { x1, y1, x2, y2, cx: (x1 + x2) / 2, cy: (y1 + y2) / 2 };
}

export interface Outline {
  points: Point[];
  closed: boolean;
}

/** An element's outline as unrotated scene-space vertices, plus closed flag. */
export function unrotatedOutline(el: ExcalidrawElement): Outline {
  const x1 = el.x;
  const y1 = el.y;
  const x2 = el.x + el.width;
  const y2 = el.y + el.height;
  const cx = (x1 + x2) / 2;
  const cy = (y1 + y2) / 2;

  switch (el.type) {
    case "diamond":
      return {
        points: [new Point(cx, y1), new Point(x2, cy), new Point(cx, y2), new Point(x1, cy)],
        closed: true,
      };
    case "line":
      return {
        points: el.points.map((p) => new Point(el.x + p[0], el.y + p[1])),
        closed: el.polygon || isPathALoop(el.points),
      };
    case "arrow":
      return { points: scenePoints(el) ?? [], closed: false };
    case "freedraw":
      return {
        points: el.points.map((p) => new Point(el.x + p[0], el.y + p[1])),
        closed: isPathALoop(el.points),
      };
    default:
      return {
        points: [new Point(x1, y1), new Point(x2, y1), new Point(x2, y2), new Point(x1, y2)],
        closed: true,
      };
  }
}

/** Bounding box of the element; when `nonRotated` is true its rotation is ignored. */
export function bounds(el: ExcalidrawElement, nonRotated = false): BoundingBox {
  const { x1, y1, x2, y2, cx, cy } = absoluteCoords(el);
  if (nonRotated || el.angle === 0) {
    return new BoundingBox(x1, y1, x2, y2);
  }
  const angle = el.angle;
  const center = new Point(cx, cy);
  const fallback = new BoundingBox(x1, y1, x2, y2);

  if (el.type === "freedraw" || el.type === "line" || el.type === "arrow") {
    const rotated = (scenePoints(el) ?? []).map((p) => p.rotated(center, angle));
    return BoundingBox.fromPoints(rotated) ?? fallback;
  }
  if (el.type === "ellipse") {
    const w = (x2 - x1) / 2;
    const h = (y2 - y1) / 2;
    const ww = Math.hypot(w * Math.cos(angle), h * Math.sin(angle));
    const hh = Math.hypot(h * Math.cos(angle), w * Math.sin(angle));
    return new BoundingBox(cx - ww, cy - hh, cx + ww, cy + hh);
  }
  const corners = unrotatedOutline(el).points.map((p) => p.rotated(center, angle));
  return BoundingBox.fromPoints(corners) ?? fallback;
}

/** Combined bounds of several elements (`getCommonBounds`); null if empty. */
export function commonBounds(elements: ExcalidrawElement[]): BoundingBox | null {
  let acc: BoundingBox | null = null;
  for (const el of elements) {
    const b = bounds(el);
    acc = acc === null ? b : acc.union(b);
  }
  return acc;
}
